src/uct/api/v2/uct_v2.h

```diff
@@ -308,6 +319,36 @@ struct uct_ep_attr {
 typedef void (*uct_md_mem_invalidate_cb_t)(void *arg);
 
 
+/**
+ * @ingroup UCT_MD
+ * @brief Operation parameters passed to @ref uct_md_mem_reg_v2.
+ */
+typedef struct uct_md_mem_reg_params {
+    /**
+     * Mask of valid fields in this structure and operation flags, using
+     * bits from @ref uct_md_mem_reg_field_mask_t. Fields not specified
+     * in this mask will be ignored. Provides ABI compatibility with respect
+     * to adding new fields.
+     */
+    uint64_t                     field_mask;
+
+    /**
+     * Operation specific flags, using bits from @ref uct_md_mem_flags_t.
+     */
+    unsigned                     flags;
+
+    /**
+     * Memory to register.
+     */
+    void                         *address;
+
+    /**
+     * Size of memory to register. Must be > 0.
+     */
+    size_t                       length;
```

Comment:
Based on the above discussion it seems like length is not optional anymore ?

Reply:
currently, it is mandatory parameter.
but we are thinking about adding Cross-GVMI support in UCX which would allow us passing rkey/rkey_buffer instead of buffer:length pair.

Reply:
@dmitrygx Can you please clarify the above. I don't understand why address and length moved to optional parameters. I will need a bit more details on this.

Reply:
> @dmitrygx Can you please clarify the above. I don't understand why address and length moved to optional parameters. I will need a bit more details on this.

@shamisp sorry missed your question...
Cross GVMI memory key is used to allow cross GVMI memory access using indirect memory registration which crosses vHCA context. (ref https://docs.nvidia.com/networking/display/ConnectX6DxFirmwarev22322004/Changes+and+New+Feature+History)
In other words, we register user-allocated virtual memory on HOST, share the key and address:length pair with another process which is running on BlueField's ARM. So, this process is able to use this memory for doing send/recv/RMA/AMO.

Reply:
@dmitrygx But I would expect that address:length pair  is still required for this registration ? Is there verbs interface for this ?

Reply:
@shamisp 

> But I would expect that address:length pair is still required for this registration ?

it is not required, since we are going to pass `rkey_buffer` which contains Cross-GVMI rkey.

> Is there verbs interface for this ?

From my knowledge, currently there are no Verbs interface for this

Reply:
@dmitrygx can provide some examples how it supposed to work ? How user will know that this is a special type of a rkey that you have to pass to another registration function ? I'm not sure I understand the workflow.

Reply:
@shamisp 

> How will user know that this is a special type of a rkey that you have to pass to another registration function ?

this is preliminary flow how Cross-GVMI feature will be implemented/used.
1. user's buffer will be registered by a process on HOST using ucp_mem_map(`SHARED`) and packing rkey_buffer.
2. sending rkey_buffer oob to a process on DPU.
3. the process on DPIU does ucp_mem_map(`rkey_buffer`).
4. the process on DPU does `ucp_am_send_nbx` providing memory handle from 3rd step.

btw, having address/length as optional parameters makes API more flexible for future updates, even if Cross-GVMI will be implemented another way,

Reply:
@dmitrygx `void *` even more flexible interface :) Our philosophy for an API - it should do one thing and do it right. 
In this specific case (remapping rkeys) I don't think this is the right interface. I think `ucp_ep_rkey_unpack` https://github.com/openucx/ucx/blob/4e1aa5e94cd56d93d6301051a82c881174aad29f/src/ucp/api/ucp.h#L2925 is probably more appropriate interface for remapping keys. Adding @yosefe to the discussion.

Reply:
> I think ucp_ep_rkey_unpack
 is probably more appropriate interface for remapping keys

yes, but we started to discuss UCP API.

sorry, I was wrong, `uct_reg_mem_reg_v2` will accept `uct_rkey_h`.

Reply:
@dmitrygx is not the unpack function directly goes to https://github.com/openucx/ucx/blob/16b610c558b0d57aeb0414c584432883749ea2fc/src/uct/ib/base/ib_md.c#L1869 

I don't think for this we need to break semantics of the `ucp_mem_map()`.

Reply:
> @dmitrygx is not the unpack function directly goes to .rkey_unpack        = uct_ib_rkey_unpack, 

we don't update neither UCP nor UCT rkey API

> I don't think for this we need to break semantics of the ucp_mem_map().

we don't do it

Reply:
I think passing `rkey_buffer`/`rkey` to `uct_mem_reg_v2()` doesn't break the idea of the registration, because:
- the result of the operation is memory handle - `uct_mem_h`
- using `rkey_buffer`/`rkey` we register memory exported by HOST to use it transprately on SmartNIC

Reply:
@dmitrygx Since you were advocating for making length/address optional I was trying to understand better the use case. At least based on the above discussion I don't think `uct_mem_reg_v2()` is the right interface for this. My preference it to move length and address back to argument list.

If you convinced rkey sharing use-case will not work with UCT interface, let's have deep dive discussion on this to understand the requirement, how it will work and how it will be used. I do believe this an interesting use case but I don't believe it will impact this specific interface.

Reply:
>  At least based on the above discussion I don't think uct_mem_reg_v2() is the right interface for this. 

Internally, it is still memory registration. IMO, from user perspective, it should look like memory registration, because we return local memory handle which could be passed to Zcopy operation.

Also, my additional points for having optional address/length (even if we will decide to not use `uct_mem_reg_v2` for passing reky_buffer):
- as an example, `uct_mem_dereg_v2` accept `memh` as optional one: https://github.com/openucx/ucx/blob/98e89780b0216e70e8c5c69efee8506152a7f238/src/uct/api/v2/uct_v2.h#L198
so, `uct_mem_dereg_v2` and `uct_mem_reg_v2` could be alligned.
- `uct_mem_reg_v2` could be extended by accepting an array of IOVs (e.g. to register under a single memory handle) instead of address/length.
- we could continue discussion of the new UCT API, while `uct_mem_reg_v2` is ready for different options.